packages/hardhat/contracts/claim/Staking.sol

```diff
@@ -0,0 +1,234 @@
+        softMfersContract.redeem(msg.sender, nftsToRedeem);
+    }
+
+    function stake(
+        uint256 amount,
+        address tokenAddress
+    ) external whenNotPaused {
+        require(amount > 0, "Amount must be greater than 0");
+        require(tokenAddress != address(0), "Invalid token address");
+
+        IERC20 token = IERC20(tokenAddress);
+        uint256 userBalance = token.balanceOf(msg.sender);
+
+        // uint256 userBalance = softToken.balanceOf(msg.sender);
+        require(userBalance >= amount, "Insufficient token balance");
+
+        uint256 userAllowance = token.allowance(msg.sender, address(this));
+        require(userAllowance >= amount, "Insufficient token allowance");
+
+        bool transferSuccess = token.transferFrom(
+            msg.sender,
+            address(this),
+            amount
+        );
+        require(transferSuccess, "Token transfer failed");
+
+        if (stakedTokens[msg.sender].amount == 0) {
+            stakedTokens[msg.sender].startTime = block.timestamp;
+        } else {
+            stakedTokens[msg.sender].totalStakingDuration +=
+                block.timestamp -
+                stakedTokens[msg.sender].stakingTime;
+        }
+
+        stakedTokens[msg.sender].amount += amount;
+        stakedTokens[msg.sender].stakingTime = block.timestamp;
+
+        emit Staked(msg.sender, amount);
+        notifyNFTContract(msg.sender);
+        adjustFees(msg.sender);
+    }
+
+    function unstake(
+        uint256 amount,
+        address tokenAddress
+    ) external whenNotPaused {
+        require(tokenAddress != address(0), "Invalid token address");
+
+        IERC20 token = IERC20(tokenAddress);
+
+        require(
+            stakedTokens[msg.sender].amount >= amount,
+            "Insufficient staked tokens"
+        );
+        uint256 stakingDuration = block.timestamp -
+            stakedTokens[msg.sender].startTime;
+        uint256 monthsStaked = (stakedTokens[msg.sender].totalStakingDuration +
+            stakingDuration) / (30 days);
+
+        uint256 penalty;
+        if (monthsStaked < 1 && stakedTokens[msg.sender].amount >= 2500e18) {
+            penalty = (amount * 1000) / 10000; // 10% penalty for early withdrawal
+        } else if (
+            monthsStaked < 2 && stakedTokens[msg.sender].amount >= 10000e18
+        ) {
+            penalty = (amount * 500) / 10000; // 5% penalty for early withdrawal
+        } else if (
+            monthsStaked < 3 && stakedTokens[msg.sender].amount >= 50000e18
+        ) {
+            penalty = (amount * 200) / 10000; // 2% penalty for early withdrawal
+        }
+
+        uint256 withdrawAmount = amount - penalty;
+        stakedTokens[msg.sender].amount -= amount;
+        stakedTokens[msg.sender].totalStakingDuration += stakingDuration;
+        stakedTokens[msg.sender].startTime = block.timestamp;
+        token.transfer(msg.sender, withdrawAmount);
+        token.transfer(address(this), penalty);
```

Comment:
Transferring the penalty amount might be superfluous here (with the current design), I'm not 100% sure. I think this is transferring the tokens owned by the staking contract, to the staking contract, so it might just be sending the tokens to itself which could be superfluous.

Once I have a unit test working, I'm sure I'll go over the rough edges and find out what we can improve.